### ISSUE TITLE
Create a nctoolbox-info function

### ISSUE THREAD

**Issue by rsignell-usgs** [nctoolbox/nctoolbox#68]
How is one expected to install and maintain nctoolbox? I've got branch 1.0.x checked out with git and am doing pulls, and this change is now merged in there. It might be nice to add a command that tells folks what branch, release, and/or version(s) they are using. It could list the java/ directory as well and maybe help with debugging and troubleshooting as well.

**Comment by drf5n** [comment]
inspired by #65.  I've pushed a version to drf5n:versioninfo that adds some Contents.m files and a nctoolbox_info.m command:

    >> help nctoolbox
      nctoolbox Tools for read-only access to Common Data Model dataset.
      Version 1.1.1rversioninfo 12-03-2015
     
      Files
        setup_nctoolbox - sets your path and classpath to access the nctoolbox functions
     
        cfdataset         - Provide access to CF/COARDS convention datasets accessable by the
        ncdataset         - Provide access to datasets accessable by the NetCDF 4 API
        ncgeodataset      - NCGEODATASET
        ncgeovariable     - Provide advanced access to variables and their related
        nctoolbox_info    - nctoolbox_info  Provide information on the nctoolbox version and dependencies
        ncugrid           - Extention of dataset object class for unstructured grid datasets.
        ncuvariable       - Extention of variable object class for unstructured grid/mesh variables.
        ncvariable        - Provide advanced access to variables and their related
     
       See https://github.com/nctoolbox/nctoolbox for more information
    
    >> nctoolbox_info
    ----------------------------------------------------------------------------------------------------
    MATLAB Version: 8.4.0.150421 (R2014b)
    MATLAB License Number: 302539
    Operating System: Mac OS X  Version: 10.10.2 Build: 14C109 
    Java Version: Java 1.7.0_55-b13 with Oracle Corporation Java HotSpot(TM) 64-Bit Server VM mixed mode
    ----------------------------------------------------------------------------------------------------
    NCTOOLBOX Tools for read-only access to Common ...    Version 1.1.1rversioninfo
    Java files in /Users/drf/matlab/nctoolbox/java
    BasicCredentialsProvider.class		je-4.0.92.jar
    BasicCredentialsProvider.java		jna-3.0.9.jar
    Contents.m				joda-time-2.1.jar
    addjars.m				log4j-1.2.17.jar
    bufr-4.3.15.jar				netcdf-4.3.15.jar
    c3p0-0.9.1.1.jar.duplicate		opendap-4.3.15.jar
    commons-httpclient-3.1.jar.duplicate	protobuf-java-2.4.1.jar
    ehcache-core-2.6.5.jar			quartz-2.1.6.jar
    grib-4.3.15.jar				setup_nctoolbox_java.m
    jcip-annotations-1.0.jar		slf4j-log4j12-1.7.2.jar
    jcl-over-slf4j-1.7.2.jar		udunits-4.3.15.jar
    jdom-2.0.2.jar.duplicate
    
    >>

**Comment by rsignell-usgs** [comment]
@drf5n , looking good!

Can you clean up this line a little (perhaps remove the rversioninfo)?
```
NCTOOLBOX Tools for read-only access to Common ...    Version 1.1.1rversioninfo
```

**Comment by drf5n** [comment]
I put that one in there specifically to identify it as my deviant variant/branch of what one would download from the official pages.    As one branches or releases,  the maintainer would need to edit the second line of  nctoolbox/Contents.m to reflect the proper version string.  

What would you like it to be? I'll push up and submit a pull request.

Does matlab have a way of doing something like 'svnversion' or 'git describe' ?   I was playing with that, but it seems very non-portable across platforms and installations.

**Comment by drf5n** [comment]
Here's the output from a modified nctoolbox_info with 'git describe; git status' included:

    nctoolbox_info
    ----------------------------------------------------------------------------------------------------
    MATLAB Version: 8.4.0.150421 (R2014b)
    MATLAB License Number: 302539
    Operating System: Mac OS X  Version: 10.10.2 Build: 14C109 
    Java Version: Java 1.7.0_55-b13 with Oracle Corporation Java HotSpot(TM) 64-Bit Server VM mixed mode
    ----------------------------------------------------------------------------------------------------
    NCTOOLBOX Tools for read-only access to Common ...    Version 1.1.1rversioninfo
    Java files in /Users/drf/matlab/nctoolbox/java
    BasicCredentialsProvider.class		je-4.0.92.jar
    BasicCredentialsProvider.java		jna-3.0.9.jar
    Contents.m				joda-time-2.1.jar
    addjars.m				log4j-1.2.17.jar
    bufr-4.3.15.jar				netcdf-4.3.15.jar
    c3p0-0.9.1.1.jar.duplicate		opendap-4.3.15.jar
    commons-httpclient-3.1.jar.duplicate	protobuf-java-2.4.1.jar
    ehcache-core-2.6.5.jar			quartz-2.1.6.jar
    grib-4.3.15.jar				setup_nctoolbox_java.m
    jcip-annotations-1.0.jar		slf4j-log4j12-1.7.2.jar
    jcl-over-slf4j-1.7.2.jar		udunits-4.3.15.jar
    jdom-2.0.2.jar.duplicate
    
    nctoolbox-20130305-20-g1c96f14
    On branch versioninfo
    Your branch is ahead of 'github/versioninfo' by 1 commit.
      (use "git push" to publish your local commits)
    
    Changes not staged for commit:
      (use "git add <file>..." to update what will be committed)
      (use "git checkout -- <file>..." to discard changes in working directory)
    
    	modified:   ../cdm/nctoolbox_info.m
    
    Untracked files:
      (use "git add <file>..." to include in what will be committed)
    
    	../Contents.m~
    	../cdm/.nctoolbox_info.m.swp
    	../cdm/ncgeovariable.m~
    	../cdm/utilities/Contents.m
    	../cdm/utilities/graphics/Contents.m
    	../demos/contrib/coawst_iso.avi
    	../demos/contrib/katama_ssh.mat
    	../demos/contrib/test_cf_ugrid3_drf.m
    	../demos/junk.ncml
    	../junk
    	../patch
    
    no changes added to commit (use "git add" and/or "git commit -a")

**Comment by hohonuuli** [comment]
I'm looking into adding a pre-commit hook to generate a text file with the branch and version info.

**Comment by drf5n** [comment]
That would be cool. Though one thing I like about using svnversion post-commit is that it can detect if you are using a modified version.  

I see that my matlab window shows the git status of the files in the working directory, so I was wondering how it can tell.

**Comment by rsignell-usgs** [comment]
We definitely can't rely on people using git, so git info should definitely be optional and handle gracefully if git is not present.

**Comment by hohonuuli** [comment]
Cool, Looks like the latest versions of Matlab have Git included: http://www.mathworks.com/help/matlab/matlab_prog/set-up-git-source-control.html?refresh=true

**Comment by hohonuuli** [comment]
In the mean time. I wrote a git hook to generate a file. The catch is you can't commit a git hook to a repo. So if you want to try this, for now:
- create a file `nctoolbox\.git\hooks\pre-commit` 
- Add the following: 

```bash
#!/usr/bin/env bash

NC_COMMIT=$(git log --pretty=format:"%h" --decorate | head -n 1)
NC_BRANCH=$(git log --pretty=format:"%d" --decorate | head -n 1)
NC_MODDATE=$(git log --pretty=format:"%cd" --decorate --date=short | head -n 1)

cat > "cdm/nctoobox-version.txt" << EOL
GIT COMMIT: $NC_COMMIT
GIT BRANCH:$NC_BRANCH
GIT DATE:   $NC_MODDATE
EOL

git add "cdm/nctoobox-version.txt"
```
- Be sure to run `chmod u+x pre-commit` on the file so it's executable.

This will create a file `cdm/nctoolbox-info.txt` and add it to the current commit. It's  contents  are like:

```
GIT COMMIT: f5cc04a
GIT BRANCH: (HEAD, origin/develop, develop)
GIT DATE:   2015-03-12

```

**Comment by drf5n** [comment]
So, branch 'master' lags branch '1.0.x' and there are a number of releases at points in between available from https://github.com/nctoolbox/nctoolbox/releases  And there's a branch 'develop'.  I'd think that the version string in nctoolbox/Contents.m should follow the releases, but I'm not sure how to think of the branches.  For example, if we added a version-containing Contents.m file to master and devel branch what should they be?

Maybe branch 1.0.x should spawn a 1.1.x or 1.x.x branch?

Also   'git describe --tags' is nice:

   1.1.1-10-g1c96f14

This is the last tagged release up through the history, the number of intervening commits, and the fingerprint for the latest commit.  I'm going to add a double plus sign to indicate that the version lags the actual commit I'll push.  As you make official releases, you should change the tag as appropriate.  Maybe the result of 'git describe --abbrev=0 --tags' would be the right text.

Hey, wouldn't the git-hook script's NC_COMMIT and NC_MODDATE also lag the commit by one?

I added the --tags option to the trapped try( git stuff) in my nctoolbox_info.  Including 'git status' is overkill, but it tells me I'm not using a clean version from github:

    ----------------------------------------------------------------------------------------------------
    MATLAB Version: 8.4.0.150421 (R2014b)
    MATLAB License Number: 302539
    Operating System: Mac OS X  Version: 10.10.2 Build: 14C109 
    Java Version: Java 1.7.0_55-b13 with Oracle Corporation Java HotSpot(TM) 64-Bit Server VM mixed mode
    ----------------------------------------------------------------------------------------------------
    NCTOOLBOX Tools for read-only access to Common ...    Version 1.1.1-10-g1c96f14++
    Java files in /Users/drf/matlab/nctoolbox/java
    BasicCredentialsProvider.class		je-4.0.92.jar
    BasicCredentialsProvider.java		jna-3.0.9.jar
    Contents.m				joda-time-2.1.jar
    addjars.m				log4j-1.2.17.jar
    bufr-4.3.15.jar				netcdf-4.3.15.jar
    c3p0-0.9.1.1.jar.duplicate		opendap-4.3.15.jar
    commons-httpclient-3.1.jar.duplicate	protobuf-java-2.4.1.jar
    ehcache-core-2.6.5.jar			quartz-2.1.6.jar
    grib-4.3.15.jar				setup_nctoolbox_java.m
    jcip-annotations-1.0.jar		slf4j-log4j12-1.7.2.jar
    jcl-over-slf4j-1.7.2.jar		udunits-4.3.15.jar
    jdom-2.0.2.jar.duplicate
    
    1.1.1-11-g8a0d23c
    On branch versioninfo
    Your branch is ahead of 'github/versioninfo' by 1 commit.
      (use "git push" to publish your local commits)
    
    Changes to be committed:
      (use "git reset HEAD <file>..." to unstage)
    
    	modified:   ../cdm/nctoolbox_info.m
    
    Untracked files:
      (use "git add <file>..." to include in what will be committed)
    
    	../Contents.m~
    	../cdm/ncgeovariable.m~
    	../cdm/utilities/Contents.m
    	../cdm/utilities/graphics/Contents.m
    	../demos/contrib/coawst_iso.avi
    	../demos/contrib/katama_ssh.mat
    	../demos/contrib/test_cf_ugrid3_drf.m
    	../demos/junk.ncml
    	../junk
    	../patch

**Comment by drf5n** [comment]
> On Mar 12, 2015, at 1:53 PM, Rich Signell <notifications@github.com> wrote:
> 
> We definitely can't rely on people using git, so git info should definitely be optional and handle gracefully if git is not present.

Even if they have git, I think the installation from a tarball bypasses all the git metadata.

**Comment by hohonuuli** [comment]
That's correct. That's why using the git hook to generate a nctoolbox-info.txt file on commits is a decent idea. The info function could read it if it exists; or try calling git from matlab if it does not.

**Comment by hohonuuli** [comment]
> Maybe branch 1.0.x should spawn a 1.1.x or 1.x.x branch?

Probably not. Releases really should be tagged in the master branch. Branches are primarily for bug fixes to that specific branch.

**Comment by drf5n** [comment]
The commit hook is decent in that it automatically maintains the cdm/nctoobox_version.txt file  (you should edit the script with s/toobox/toolbox/) but with a pre-commit hook it is a bit confusing in that the documented state will always lag the actual state by one commit.  Would a post-commit update be better? 

The most recent tags on the master, 1.0.x, develop, and my versioninfo  branch are:

    $ git describe --tags master 1.0.x develop versioninfo 
    nctoolbox-20140414-beta-10-g3161fee
    1.1.1-7-g96de0bf
    nctoolbox-20140414-beta-31-g0e9f745
    1.1.1-13-g8582810

I'm not sure what I did to install last, but I think it might have been: 
   git clone  git@github.com:nctoolbox/nctoolbox.git  # since this was the default
   git checkout 1.1.1    # since this looked like the latest release
   git checkout 1.0.x    # since I got the detached head message

I wanted to use something fairly stable on these machines, but that worked.  And I wanted to be able to recommend it to the students in the class.

**Comment by hohonuuli** [comment]
> Would a post-commit update be better? 

It's a good point about the version lag. The problem with a post commit hook is that you can't add the current version of the nctoolbox-version.txt file to the commit. Hmmmm. Must think on this.

**Comment by hohonuuli** [comment]
Just a note that there's been a number of changes on the 1.0.x branch. Unless anyone objects; I'm pulling the changes back into the develop branch.

**Comment by rsignell-usgs** [comment]
Sounds good to me.

**Comment by hohonuuli** [comment]
Merged. See 3c86be0

**Comment by drf5n** [comment]
Because of the lag problem it is hard to put the source control state into the controlled source.   On some SVN-using software I work with, I put SVNVERSION=`svnversion` in the makefile so I can compile the version info into the code.  For Git, I think the `git describe --tags` would be equivalent, but both require access to the source control software on the client side, and neither would work here with tarball installations.   Because of that, I think the Contents.m Version string should probably be manually updated to reflect the releases and shouldn't try to capture the full, detailed state of the source control.

**Comment by drf5n** [comment]
> Probably not. Releases really should be tagged in the master branch. Branches are primarily for bug fixes to that specific branch.

Shouldn't the master branch then cover at least 1.1.1?   If it is supposed to be lagging the releases, maybe there should be a branch like Debian's 'stable' or 'oldstable' that follows the development more conservatively?

**Comment by hohonuuli** [comment]
Yes, the master branch should cover 1.1.1. I merged the 1.0.x branch (which includes the 1.1.1 tag) back into the master branch. I'll try to run tests later today and retag the master branch with 1.1.2. The develop branch has been tagged with _2.0-SNAPSHOT_.

I also monkeyed around with `nctoolbox_info` to fall back to using the file our hooks are generating if git isn't available, but only on the develop branch (See https://github.com/nctoolbox/nctoolbox/blob/develop/cdm/nctoolbox_info.m).

**Comment by hohonuuli** [comment]
Just a note that that I changed the output of nctoolbox_info slightly:

```
----------------------------------------------------------------------------------------------------
MATLAB Version: 8.5.0.197613 (R2015a)
MATLAB License Number: 177731
Operating System: Mac OS X  Version: 10.10.2 Build: 14C1510 
Java Version: Java 1.7.0_60-b19 with Oracle Corporation Java HotSpot(TM) 64-Bit Server VM mixed mode
----------------------------------------------------------------------------------------------------
MATLAB                                                Version 8.5        (R2015a)
CDM Common Data Model tools from NCTOOLBOX            Version 2.0-SNAPSHOT       
Statistics and Machine Learning Toolbox               Version 10.0       (R2015a)

Java jar files included with nctoolbox:
	/Users/brian/workspace/nctoolbox/cdm/bufr-4.3.23.jar
	/Users/brian/workspace/nctoolbox/cdm/commons-codec-1.2.jar
	/Users/brian/workspace/nctoolbox/cdm/ehcache-core-2.6.2.jar
	/Users/brian/workspace/nctoolbox/cdm/grib-4.3.23.jar
	/Users/brian/workspace/nctoolbox/cdm/jcip-annotations-1.0.jar
	/Users/brian/workspace/nctoolbox/cdm/jcl-over-slf4j-1.7.5.jar
	/Users/brian/workspace/nctoolbox/cdm/jdom2-2.0.4.jar
	/Users/brian/workspace/nctoolbox/cdm/je-4.0.92.jar
	/Users/brian/workspace/nctoolbox/cdm/jna-3.0.9.jar
	/Users/brian/workspace/nctoolbox/cdm/joda-time-2.2.jar
	/Users/brian/workspace/nctoolbox/cdm/jsoup-1.7.2.jar
	/Users/brian/workspace/nctoolbox/cdm/netcdf-4.3.23.jar
	/Users/brian/workspace/nctoolbox/cdm/opendap-4.3.23.jar
	/Users/brian/workspace/nctoolbox/cdm/protobuf-java-2.5.0.jar
	/Users/brian/workspace/nctoolbox/cdm/quartz-2.1.1.jar
	/Users/brian/workspace/nctoolbox/cdm/slf4j-api-1.7.7.jar
	/Users/brian/workspace/nctoolbox/cdm/slf4j-log4j12-1.7.7.jar
	/Users/brian/workspace/nctoolbox/cdm/udunits-4.3.23.jar

Java class files included with nctoolbox:
	/Users/brian/workspace/nctoolbox/cdm/BasicCredentialsProvider.class

nctoolbox revision in /Users/brian/workspace/nctoolbox:
GIT COMMIT:   dbf1798
GIT BRANCH:   develop
GIT DESCRIBE: 2.0-SNAPSHOT-1-gdbf1798
```

**Comment by drf5n** [comment]
Looks good.  

If I understand the branch naming correctly, users should be using master, and it should work as it described, '1.0.x' is a transient, issue-fixing development branch, and 'develop' is a longer term unstable development branch.  I'll point the recent installs to 'master'

**Comment by rsignell-usgs** [comment]
When I branched `1.0.x` off of `master`, I didn't use most recent version of `master` because it was broken.   
Did `master` subsqeuently get fixed by @hohonuuli?

**Comment by hohonuuli** [comment]
I just merged all the 1.0.x changes back into master, so the fixes _should_ be there. I've got a few tasks to tackle, but I plan to run tests on master and verify the merges are correct today. If everything looks good, I'll tag a 1.1.2 release.

**Comment by hohonuuli** [comment]
@drf5n Yes, your description of the branches is spot on to what I'm thinking.

**Comment by hohonuuli** [comment]
OK, To help us out a little there's a new branch called `integration` where you can do fixes and such and test them without putting them on the `master` branch until they are ready. I threw a brief note about branches at https://github.com/nctoolbox/nctoolbox/wiki/Information-for-Developers

Also, I've finished the merge between `1.0.x` into `master`. There are a few fixes in ncdataset that were in master but not in 1.0.x. You may want to test it out and let me know if it's working for you AOK. runtests all passes in Matlab R2014b. Once I get the AOK, I'll tag a new release.

**Comment by hohonuuli** [comment]
Just a note that `runtests` also passes in Matlab 2015a.

**Comment by hohonuuli** [comment]
p.s. I don't know what your development set up is but if you're doing merging it's very handy to configure a graphical diff tool to work with git via `git difftool`. I'm on a mac and I use [Kaleidoscipe](http://www.kaleidoscopeapp.com/) which is a bit pricy but kind of awesome. There's cheaper/free options around too. I can diff and merge edits between changesets using commands like `git difftool HEAD 1.0.x`

**Comment by drf5n** [comment]
Thanks for the developer info.  Master looks like the one to track.  

One thing that looks confusing on https://github.com/nctoolbox/nctoolbox/releases/tag/2.0-SNAPSHOT is the tag-moving language.  I'd think tags shouldn't change--they are essentially aliases for commits in stored in .git/refs/tags/*   Moving or changing public tags could potentially make the same tag point to two different histories. Branches seem the right tool for evolving code histories.

My environment is typically emacs (Aquamacs on the mac) and command-line git.  I sometimes use gitk if there's a confusing history.  For myself, I often like 'git cherry-pick' over merge

**Comment by hohonuuli** [comment]
A tag is just a way to quickly id and reference a change set. For the most part, we won't be moving tags. I added this particular one because `git describe` on the develop branch dumps out a tag like 'nctoolbox-20141012`, which in our new version scheme is confusing. I added the tag `2.0-SNAPSHOT` so nctoolbox_info would print out something that made more sense. However github generates releases from tags, and a SNAPSHOT by definition is changing. I may never move that tag, but I don't want to give users a feeling that they can rely on the contents of the 2.0-SNAPSHOT tag either. But also, I have no issue with moving a SNAPSHOT tag around though. Git certainly allows it.

**Comment by drf5n** [comment]
Git allows you to move your own tags around, but it wont move other peoples.  From 'git help tag'

       If you never pushed anything out, just re-tag it. Use "-f" to replace the
       old one. And you're done.

       But if you have pushed things out (or others could just read your
       repository directly), then others will have already seen the old tag. In
       that case you can do one of two things:

        1. The sane thing. Just admit you screwed up, and use a different name.
           Others have already seen one tag-name, and if you keep the same name,
           you may be in the situation that two people both have "version X", but
           they actually have different "X"'s. So just call it "X.1" and be done
           with it.

        2. The insane thing. You really want to call the new version "X" too,
           even though others have already seen the old one. So just use git tag
           -f again, as if you hadn't already published the old one.

       However, Git does not (and it should not) change tags behind users back.
       So if somebody already got the old tag, doing a git pull on your tree
       shouldn't just make them overwrite the old one.
       ....

**Comment by hohonuuli** [comment]
Ahh, good to know. I take it back, I won't move the tag.

**Comment by drf5n** [comment]
Maybe the version info hook would be better as a post-checkout hook, so its data matches the repository state and does not lag by one commit.  The lagged info is confusing when working on branches, and since the cdm/nctoolbox-version.txt is under source control, it can cause extra commits and conflicts for insignificant local/offline revisions of history.  For folks using non-git-managed released tarball installations, maybe the nctoolbox info could report a manually hard-coded release tag unless there is a git post-checkout-generated cdm/nctoolbox-info-co.txt file.     

Since you can't commit the .git/hooks/* files, if those aren't installed, chmod a+x, and maintained by all the committers, the cdm/nctoolbox-version.txt file may not stay synchronized as expected.
